### ISSUE TITLE
consdier deprecate connection.execute("plain string"), supercede with connection.exec_driversql("plain string")

### ISSUE THREAD

**Issue by jtrakk** [sqlalchemy/sqlalchemy#4848]
A [common error](https://stackoverflow.com/questions/8657508/strange-sqlalchemy-error-message-typeerror-dict-object-does-not-support-inde) seems to be related to using `%` without `text()`. If it is possible to detect this situation and report a helpful error message, that would be helpful.

**Comment by zzzeek** [comment]
this is a driver-level error, and it is not necessary to double percent signs for all drivers (usually just psycopg2 and some of the mysql drivers) nor is it an error to pass plain strings to execute(), as the usage contract there is that you are talking to the DBAPI directly.   on the SQLAlchemy side, removing this ability altogether is something I might consider for SQLAlchemy 2.0 but short of that there's no generalized way to improve this error message locally; you should report this to the drivers such as psycopg2 etc.

**Comment by zzzeek** [comment]
twitter thread at https://twitter.com/zzzeek/status/1170352961995886593 has my thoughts:

* `connection.execute(text("asdf"))` is how to invoke string SQL
* `connection.exec_driversql("plain text")` is the "use the driver directly" version
* `connection.execute("string") emits deprecation warnings
* no short term plans to auto-coerce "string" into text("string"), trying to get away from auto-coercion
* huge inconsistency that session.execute("string") **DOES**  do the implicit coercion, so will deprecate the implicit coercion there as well

Basically along with the recent changes to order_by("str"), group_by("str"), if you are writing textual SQL in SQLAlchemy anywhere you should be explicitly using a text() constructor, familiarizing with all the things it can do, and that's how it will be .  explicit everywhere.  that's the proposal, anyway

**Comment by zzzeek** [comment]
a proposal for a modification based on some twitter feedback:

* connection.execute("text") will coerce to text() in SQLAlchemy 2.0.  The deprecation warning simply says, "this will be treated as text() in 2.0, not as driver SQL.  for forwards compatibility, use text("string") or exec_driversql("text").    this allows for there to be an upgrade path while still maintaining that connection.execute("sql") can be used.  

I'm not 100% sure about this, however, there is the balance between "why should it be connection.execute(text("sql"))" and the "there should be only one way to do it" answer.

**Comment by CaselIT** [comment]
Since one of the objectives of 2 is to remove the "magic" I think that we should avoid the implicit coercion of text in execute and raise an error instead.
The coercion should happen in 1.4 with a deprecation warning.

After a plan is decided I can try this one

**Comment by zzzeek** [comment]
I proposed that up above (e.g. text() is required), some folks on twitter (see thread above) were a little uncomfortable with it.   if you extrapolate that "tweets by people who follow me" count for thousands of users who just use the product then changing it that dramatically might reduce adoption.    

it's not clear on how to make decisions like these since yes, you have to provide text() is more explicit but then again, if someone says conn.execute("Stirng"), the intent should be very clear.    I can go either way.

**Comment by zzzeek** [comment]
the only confusion about the intent of connection.execute("Some string") is about things like escaping and bound parameter formats.   that would be changing here.  so in that regard, making it require text() is the better option anyway.   but it would affect session.execute() also.

**Comment by CaselIT** [comment]
I'll have a look at the twitter thread. 
I think that in this case explicit is better than implicit. Maybe the coercion+deprecation could be kept for longer, meaning that v2 would still have it, but I don't think that an implicit coercion would be the way to go, because you would have issues with "why execute transforms my sql string to text?".

Agree that the behavior should be the same with engine (that should go away), connection and session.

**Comment by CaselIT** [comment]
btw twitter is the worst wrt threads:

this is better https://threadreaderapp.com/thread/1170352961995886593.html?refreshed=yes

**Comment by zzzeek** [comment]
I use tweetdeck usually

**Comment by CaselIT** [comment]
Well regardless at what happens next I think the coercion of str to text with the deprecation and the new method could be implemented.

Regarding the new method `exec_driversql` is the preferred option? Alternative maybe are `exec_driver_sql`, `execute_driver_sql` (too long?), `execute_sql` (not really true, since it could be changed by the driver)

Personally I'm for `exec_driver_sql` or `execute_driver_sql`

**Comment by zzzeek** [comment]
exec_driver_sql   it's long enough as it is

**Comment by CaselIT** [comment]
I've started working on this.

What should `conn.scalar("select 1;")` do?
Currently it passes to execute, but since we are warning now for execute with a string, should it transform to text or use exec_driver_sql when the input is a string?

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **master** branch:

**Coerce string in execute to :func:`.text` and introduce `exec_driver_sql`** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1753

**Comment by CaselIT** [comment]
Ok, the status of the gerrit is that it should be ready for a first review.
The `keystone` tests will inevitably fail, since it fails on warning and it uses an `execute` with a string in all test setup.

**Comment by zzzeek** [comment]
OK I can propose patches for keystone

**Comment by CaselIT** [comment]
Now that I think about it, I think I've left the normal warning in the current patch.
Maybe keystone already ignores the removed in 2 warnings?

**Comment by zzzeek** [comment]
"yes" and "not yet".  i can get keystone to ignore the removed in 2.0 warnings if you change the patch.

**Comment by CaselIT** [comment]
Ok, I've updated the patch
